signal_op/extract_events.m

```diff
@@ -0,0 +1,174 @@
+function event_times = extract_events(trace,varargin)
+% event_times = extract_events(trace,varargin)
+%
+% Extract event times from the ICA trace of a cell doing 2 iterations
+%
+%   trace : ica trace of a single cell 
+%   An additional inout argument set to 1 will generate plots 
+% 
+%   event_times : [# of events] x 1 array 
+%
+% Hakan Inan (Mar 15)
+%
+
+%%%%
+%some parameters that can be changed within the script (advanced):
+%%%%
+%mad for df/f trace is multiplied with below to get the coarse threshold
+coarse_thresh_multiplier = 9;
+%mad for deconvolved trace is multiplied with below to get the fine threshold
+fine_thresh_multiplier = 15;
+%offset of the impulse for deconvolution
+offset = 0;
+
+% Length before and after the maximum of the characteristic event, h
+num_frames_before_max = 20;
+num_frames_after_max = 100;
+
+
+idx_event_frames = extract_candidate_events(trace,coarse_thresh_multiplier,0);
+out = iterate_event_indices(idx_event_frames);
+idx_event_frames_v2 = out{1};h_1stround = out{2};
+out = iterate_event_indices(idx_event_frames_v2);
+idx_event_frames_v3 = out{1};h_2ndround = out{2};g_2ndround = out{3};
+event_times = idx_event_frames_v3;
+
+
+%%%%%%%%%%%%%%
+%visualization
+%%%%%%%%%%%%%%
+if ~isempty(varargin)
+    if varargin{1}==1
+        figure,
+        plot(h_1stround,'k');
+        hold on
+        plot(h_2ndround,'r');
+        hold off
+        legend('the characteristic event, 1st iteration','the characteristic event, 2nd iteration');
+        xlabel('frame')
+
+        trace = trace / max(trace);
+        mad_scale_coarse = compute_mad(trace);
+        threshold_coarse = coarse_thresh_multiplier*mad_scale_coarse;
+
+        filtered =  conv(trace,g_2ndround,'same');
+        filtered = filtered / max(filtered);
+        mad_scale_fine = compute_mad(filtered);
+        threshold_fine = fine_thresh_multiplier*mad_scale_fine;
+
+        spikes_vec_fine = zeros(1,length(trace));
+        spikes_vec_fine(idx_event_frames_v3) = 1;
+        spikes_vec_coarse = zeros(1,length(trace));
+        spikes_vec_coarse(idx_event_frames) = 1;
+
+        figure,
+        plot(trace,'k');
+        hold on
+        plot(filtered,'m');
+        plot(1:length(trace),repmat(threshold_coarse,1,length(trace)),'--b');%coarse threshold
+        plot(1:length(trace),repmat(threshold_fine,1,length(trace)),'--r');%fine threshold
+        stem(spikes_vec_fine*threshold_fine);
+        stem(spikes_vec_coarse*threshold_coarse);
+        hold off
+        xlabel('frame')
+        ylabel('a.u.')
+        title('raw and deconvolved ica traces together with the thresholds and the detected events')
+        legend('raw trace','deconvolved trace','coarse threshold','deconv threshold')
+    end
+end
+
+%%%%%%%%%%%%%%%%%%%
+%internal functions
+%%%%%%%%%%%%%%%%%%%
+function out = iterate_event_indices(idx_event_frames_in)
+
+    h_len = num_frames_before_max+num_frames_after_max+1;
+    %surrogate waveform
```

Comment:
Is "surrogate waveform" referring to `event_mat`, which seems to be aligned waveforms for each event?